File: src/2025-02-27-deepclaude37-rust/go/exercises/practice/crypto-square/crypto_square.go
package cryptosquare

import (
	"math"
	"strings"
	"unicode"
)

func Encode(pt string) string {
	// Normalize the input text
	var normalized strings.Builder
	for _, ch := range pt {
		if unicode.IsLetter(ch) || unicode.IsDigit(ch) {
			normalized.WriteRune(unicode.ToLower(ch))
		}
	}
	normalizedStr := normalized.String()
	length := normalized.Len()
	if length == 0 {
		return ""
	}

	// Calculate rectangle dimensions
	c := int(math.Ceil(math.Sqrt(float64(length))))
	r := int(math.Ceil(float64(length) / float64(c)))

	// Pad the normalized string to fill the rectangle
	requiredLength := r * c
	pad := requiredLength - length
	normalizedStr += strings.Repeat(" ", pad)

	// Split into rows
	rows := make([]string, r)
	for i := 0; i < r; i++ {
		start := i * c
		end := start + c
		rows[i] = normalizedStr[start:end]
	}

	// Read columns to form the encoded string
	var encoded strings.Builder
	for col := 0; col < c; col++ {
		for _, row := range rows {
			encoded.WriteByte(row[col])
		}
	}
	encodedStr := encoded.String()

	// Split into chunks and join with spaces
	var result strings.Builder
	for i := 0; i < c; i++ {
		start := i * r
		end := start + r
		if i > 0 {
			result.WriteByte(' ')
		}
		result.WriteString(encodedStr[start:end])
	}

	return result.String()
}

File: src/2025-02-27-deepclaude37-rust/cpp/exercises/practice/crypto-square/crypto_square.cpp
#include "crypto_square.h"
#include <cmath>
#include <cctype>
#include <string>

namespace crypto_square {

cipher::cipher(const std::string& plaintext) {
    // Normalize text: remove spaces and punctuation, convert to lowercase
    for (char ch : plaintext) {
        if (std::isalnum(static_cast<unsigned char>(ch))) {
            normalized_ += std::tolower(static_cast<unsigned char>(ch));
        }
    }
    
    size_t len = normalized_.length();
    if (len == 0) {
        columns_ = 0;
        rows_ = 0;
        return;
    }
    
    // Calculate rectangle dimensions
    columns_ = static_cast<size_t>(std::ceil(std::sqrt(len)));
    rows_ = (len + columns_ - 1) / columns_; // ceiling division
}

std::string cipher::normalized_cipher_text() const {
    if (normalized_.empty()) {
        return "";
    }
    
    size_t len = normalized_.length();
    
    // Pad normalized text to fill rectangle
    std::string paddedText = normalized_;
    paddedText.resize(rows_ * columns_, ' ');
    
    // Build encoded string by reading columns
    std::string encoded;
    for (size_t col = 0; col < columns_; ++col) {
        for (size_t row = 0; row < rows_; ++row) {
            size_t index = row * columns_ + col;
            encoded += paddedText[index];
        }
    }
    
    // Split into chunks and join with spaces
    std::string result;
    for (size_t i = 0; i < columns_; ++i) {
        if (i > 0) {
            result += ' ';
        }
        result += encoded.substr(i * rows_, rows_);
    }
    
    return result;
}

}  // namespace crypto_square
